### ISSUE TITLE
`clippy::double_ended_iterator_last` is dangerous to autoapply

### ISSUE THREAD

**Issue by GoldsteinE** [rust-lang/rust-clippy#14444]
### Summary

Clippy suggests changing `.last()` to `.next_back()` which changes semantics when iterator pipeline is impure.

### Reproducer

Consider this (a bit contrived) code:

```rust
let mut squares = vec![];
let last_square = [1, 2, 3]
    .into_iter()
    .map(|x| {
        squares.push(x * x);
        Some(x * x)
    })
    .last();
dbg!(squares, last_square);
```

Clippy suggests changing `.last()` to `.next_back()` which dramatically changes the behaviour.

### Version

```text
rustc 1.87.0-nightly (3ea711f17 2025-03-09)
binary: rustc
commit-hash: 3ea711f17e3946ac3f4df11691584e2c56b4b0cf
commit-date: 2025-03-09
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-bug

**Comment by GoldsteinE** [comment]
I think that (if this lint is to remain warn-to-default):
- It shouldn't be `MachineApplicable`.
- The warning message should clearly state that this may change semantics.
